hydroflow/src/builder/mod.rs

```diff
@@ -0,0 +1,244 @@
+//! Hydroflow's Surface API.
```

Comment:
RENDERED:

Hydroflow's Surface API.

## Internal Documentation

Due to the limitations of type-level programming in Rust, this code is
"baklava" code containing lot of layers. Each layer does one thing, then
constructs the next layer(s) down. This table describes what each layer
does and is named. Layers are listed starting from the highest
(user-facing API) layer and ending with the lowest (code-running) layer.

### Layer Structure
```text
              (A) Surface API
        (B) (Push Surface Reversed*)
                 /            \
                /              \
       (C) Connector   (D) Subgraph Builder
                                 |
                      (E) Iterator/Pusherator
```
<sup>*Only used with `Push` to reverse the ownership direction.</sup>

### Layer Descriptions
<table>
<tr>
    <th rowspan="2">Layer</th>
    <th rowspan="2">Purpose</th>
    <th colspan="2">Naming</th>
</tr>
<tr>
   <th>Pull</th>
   <th>Push</th>
</tr>
<tr>
    <td>(A) The Surface API</td>
    <td rowspan="2">
        &bull; Presents a clean functional-chaining API for users.<br>
        &bull; Consumed to simultaneously create a (C) connector and (D) builder.<br>
        &bull; <strong>Push Only</strong>: Extra layer needed to reverse ownership order.
    </td>
    <td><code>PullSurface</code></td>
    <td><code>PushSurface</code></td>
</tr>
<tr>
    <td>(B) Push Surface Reversed</td>
    <td>N/A</td>
    <td><code>PushSurfaceReversed</code></td>
</tr>
<tr>
    <td>(C) Connectors</td>
    <td>
        &bull; Connects <code>OutputPort</code>s and <code>InputPort</code>s, splits type lists in order to do so.<br>
        &bull; Does not go to any lower layers.<br>
        &bull; Uses the input/output <code>HandoffList</code> variadic type.
    </td>
    <td><code>PullConnect</code></td>
    <td><code>PushConnect</code></td>
</tr>
<tr>
    <td>(D) Subgraph Builders</td>
    <td>
        &bull; On each subgraph invocation, constructs the (E) iterators and pivot which will be run.<br>
        &bull; Is owned by the subgraph task, lends closures to (E) iterators.<br>
        &bull; Uses the input/output <code>HandoffList</code> variadic type.
    </td>
    <td><code>PullBuild</code></td>
    <td><code>PushBuild</code></td>
</tr>
<tr>
    <td>(E) Iterators</td>
    <td>
        &bull; Runs code on individual dataflow elements, in the case of dataflow.<br>
        &bull; In the future, will correspond to semilattice morphisms alternatively.
    </td>
    <td><code>std::iter::Iterator</code></td>
    <td><code>crate::compiled::Pusherator</code></td>
</tr>
</table>

### How the code actually runs

The layers are used in [HydroflowBuilder::add_subgraph]. The method
receives a pivot with `PullSurface` and `PushSurfaceReversed` halves. Then
`into_parts()` splits them into the build half and the connector half.
The build half is used to create the subgraph which gives back
input/output ports. Then the connector half is used to connect up those
ports.